src/StoreApi/Utilities/OrderController.php

```diff
@@ -353,56 +375,33 @@ protected function validate_allowed_country( $country, array $allowed_countries
 	/**
 	 * Check all required address fields are set and return errors if not.
 	 *
-	 * @param array     $address Address array.
+	 * @param \WC_Order $order Order object.
 	 * @param string    $address_type billing or shipping address, used in error messages.
 	 * @param \WP_Error $errors Error object.
 	 */
-	protected function validate_address_fields( $address, $address_type, \WP_Error $errors ) {
+	protected function validate_address_fields( \WC_Order $order, $address_type, \WP_Error $errors ) {
 		$all_locales    = wc()->countries->get_country_locale();
+		$address        = $order->get_address( $address_type );
 		$current_locale = isset( $all_locales[ $address['country'] ] ) ? $all_locales[ $address['country'] ] : [];
 
-		/**
-		 * We are not using wc()->counties->get_default_address_fields() here because that is filtered. Instead, this array
-		 * is based on assets/js/base/components/cart-checkout/address-form/default-address-fields.js
-		 */
-		$address_fields = [
-			'first_name' => [
-				'label'    => __( 'First name', 'woo-gutenberg-products-block' ),
-				'required' => true,
-			],
-			'last_name'  => [
-				'label'    => __( 'Last name', 'woo-gutenberg-products-block' ),
-				'required' => true,
-			],
-			'company'    => [
-				'label'    => __( 'Company', 'woo-gutenberg-products-block' ),
-				'required' => false,
-			],
-			'address_1'  => [
-				'label'    => __( 'Address', 'woo-gutenberg-products-block' ),
-				'required' => true,
-			],
-			'address_2'  => [
-				'label'    => __( 'Apartment, suite, etc.', 'woo-gutenberg-products-block' ),
-				'required' => false,
-			],
-			'country'    => [
-				'label'    => __( 'Country/Region', 'woo-gutenberg-products-block' ),
-				'required' => true,
-			],
-			'city'       => [
-				'label'    => __( 'City', 'woo-gutenberg-products-block' ),
-				'required' => true,
-			],
-			'state'      => [
-				'label'    => __( 'State/County', 'woo-gutenberg-products-block' ),
-				'required' => true,
-			],
-			'postcode'   => [
-				'label'    => __( 'Postal code', 'woo-gutenberg-products-block' ),
-				'required' => true,
-			],
-		];
+		$additional_fields = $this->additional_fields_controller->get_all_fields_from_order( $order );
+
+		foreach ( $additional_fields as $field_id => $field_value ) {
+			$prefix = '/' . $address_type . '/';
+			if ( strpos( $field_id, $prefix ) === 0 ) {
+				$address[ str_replace( $prefix, '', $field_id ) ] = $field_value;
+			}
+		}
+
+		$fields              = $this->additional_fields_controller->get_additional_fields();
+		$address_fields_keys = $this->additional_fields_controller->get_address_fields_keys();
+		$address_fields      = array_filter(
+			$fields,
+			function( $key ) use ( $address_fields_keys ) {
+				return in_array( $key, $address_fields_keys, true );
+			},
+			ARRAY_FILTER_USE_KEY
+		);
```

Comment:
We also validate fields and key sure required fields are set, we return a set of errors alongside other missing fields.